crates/sqlbuiltins/src/functions/scalars/mod.rs

```diff
@@ -62,17 +64,11 @@ fn get_nth_scalar_value(
     match input.get(n) {
         Some(input) => match input {
             ColumnarValue::Scalar(scalar) => Ok(ColumnarValue::Scalar(op(scalar.clone())?)),
-            ColumnarValue::Array(arr) => {
-                let mut values = Vec::with_capacity(arr.len());
-
-                for idx in 0..arr.len() {
-                    values.push(op(ScalarValue::try_from_array(arr, idx)?)?);
-                }
-
-                Ok(ColumnarValue::Array(ScalarValue::iter_to_array(
-                    values.into_iter(),
-                )?))
-            }
+            ColumnarValue::Array(arr) => Ok(ColumnarValue::Array(scalar_iter_to_array(
+                (0..arr.len()).map(|idx| -> Result<ScalarValue, ExtensionError> {
+                    Ok(op(ScalarValue::try_from_array(arr, idx)?)?)
```

Comment:
I'd prefer to avoid using `try_from_array`. There's quite a bit of work being done in there that'll slow things down in a hot loop like this. 

I'd suggest that we pass in the return type to the `get_nth` function. Then we can try to extract the value directly, skipping all of the checks that happen in `try_from_array`. Similarly, `scalar_iter_to_array` could skip peeking at the elements if it knew the return type.